contracts/ics/chan/src/entry.rs

```diff
@@ -0,0 +1,125 @@
+use crate::error::{Error, Result};
+use alloc::vec::Vec;
+use ckb_ics_axon::handler::{
+    handle_channel_open_ack_and_confirm, handle_channel_open_init_and_try, Client, IbcChannel,
+    IbcConnections,
+};
+use ckb_ics_axon::message::{
+    Envelope, MsgChannelOpenAck, MsgChannelOpenConfirm, MsgChannelOpenInit, MsgChannelOpenTry,
+    MsgType,
+};
+use ckb_standalone_types::prelude::Entity;
+#[cfg(feature = "debugging")]
+use ckb_std::ckb_types::prelude::*;
+use ckb_std::error::SysError;
+use ckb_std::{ckb_constants::Source, high_level as hl};
+use rlp::decode;
+use tiny_keccak::{Hasher as _, Keccak};
+
+pub fn main() -> Result<()> {
+    let envelope = load_envelope()?;
+    match envelope.msg_type {
+        MsgType::MsgChannelOpenInit
+        | MsgType::MsgChannelOpenTry
+        | MsgType::MsgChannelOpenAck
+        | MsgType::MsgChannelOpenConfirm => {}
+        _ => return Ok(()),
+    }
+
+    let client_data = hl::load_cell_data(0, Source::CellDep).map_err(|_| Error::LoadCellDataErr)?;
+    let client = decode::<Client>(&client_data).map_err(|e| Error::Encoding)?;
+
+    let input_data = hl::load_cell_data(0, Source::GroupInput)?;
+    let expected_input_hash: [u8; 32] = input_data
+        .try_into()
+        .map_err(|e| Error::ChannelHashUnmatch)?;
+
+    let output_data = hl::load_cell_data(0, Source::GroupOutput)?;
+    let expected_output_hash: [u8; 32] = output_data
+        .try_into()
+        .map_err(|e| Error::ChannelHashUnmatch)?;
+
+    let witness_args0 = hl::load_witness_args(0, Source::Input)?;
+    let witness_args1 = hl::load_witness_args(1, Source::Output)?;
+
+    let old_cell_data = witness_args0.input_type();
```

Comment:
what do you think of using `witnesses[0].input_type` for old data and `witnesses[1].output_type` for new data, I'm wondering what exactly the data is in `witnesses[0].output_type` and `witnesses[1].input_type`